ui/job-view/KeyboardShortcuts.jsx

```diff
@@ -0,0 +1,354 @@
+
+    if (selectedJob) {
+      this.$rootScope.$emit(thEvents.addRelatedBug, selectedJob);
+      $('#related-bug-input').focus();
+    }
+  }
+
+  // pin selected job to pinboard and enter classification
+  pinEditComment() {
+    const { selectedJob } = this.$rootScope;
+
+    if (selectedJob) {
+      this.$rootScope.$emit(thEvents.jobPin, selectedJob);
+      $('#classification-comment').focus();
+    }
+  }
+
+  // clear the PinBoard
+  clearPinboard() {
+    this.$rootScope.$emit(thEvents.clearPinboard);
+  }
+
+  saveClassification() {
+    this.$rootScope.$emit(thEvents.saveClassification);
+  }
+
+  // delete classification and related bugs
+  deleteClassification() {
+    const { selectedJob } = this.props;
+
+    if (selectedJob) {
+      this.$rootScope.$emit(thEvents.deleteClassification);
+    }
+  }
+
+  // open the logviewer for the selected job
+  openLogviewer() {
+    this.$rootScope.$emit(thEvents.openLogviewer);
+  }
+
+  // retrigger selected job
+  jobRetrigger() {
+    const { selectedJob } = this.props;
+
+    if (selectedJob) {
+      this.$rootScope.$emit(thEvents.jobRetrigger, selectedJob);
+    }
+  }
+
+  // select next job tab
+  selectNextTab() {
+    const { selectedJob } = this.props;
+
+    if (selectedJob) {
+      this.$rootScope.$emit(thEvents.selectNextTab);
+    }
+  }
+
+  /**
+  * Filter and Help Shortcuts
+  */
+
+  // enter a quick filter
+  quickFilter() {
+    $('#quick-filter').focus();
+  }
+
+  // clear the quick filter field
+  clearFilter() {
+    const { filterModel } = this.props;
+
+    filterModel.removeFilter('searchStr');
+  }
+
+  toggleOnScreenShortcuts() {
+    this.$rootScope.setOnscreenShortcutsShowing(!this.$rootScope.onscreenShortcutsShowing);
+    this.$rootScope.$apply();
+  }
+
+  /**
+   * Autoclassify Shortcuts
+   */
+
+  // ignore selected in the autoclasify panel
+  autoclassifySelectOption(ev) {
+    this.$rootScope.$emit(
+      thEvents.autoclassifySelectOption,
+      ev.key === 'o' ? 'manual' : ev.key);
+  }
+
+  // select all remaining unverified lines on the current job
+  autoclassifyChangeSelection() {
+    this.$rootScope.$emit(
+      thEvents.autoclassifyChangeSelection,
+      'all_next',
+      false);
+  }
+
+  // toggle edit mode for selected lines
+  autoclassifyToggleEdit() {
+    const { selectedJob } = this.$rootScope;
+
+    if (selectedJob) {
+      this.$rootScope.$emit(thEvents.autoclassifyToggleEdit);
+    }
+  }
+
+  // ignore selected in the autoclasify panel
+  autoclassifyIgnore() {
+    this.$rootScope.$emit(thEvents.autoclassifyIgnore);
+  }
+
+  // save all in the autoclasify panel
+  autoclassifySaveAll() {
+    this.$rootScope.$emit(thEvents.autoclassifySaveAll);
+  }
+
+  // toggle more/fewer options in the autoclassify panel
+  autoclassifyToggleExpandOptions() {
+    this.$rootScope.$emit(thEvents.autoclassifyToggleExpandOptions);
+  }
+
+  // select previous unverified log line
+  previousAutoclassifyOption(ev) {
```

Comment:
Since many of these methods are structured the same, what do you think about making some of them generic and having the handler keys pass in arguments?

Reply:
It's true, though this is the first step in making these call functions in a ``selectedJob`` context API coming in a PR very soon.  This will get rid of a lot of the ``this.$rootScope.$emit`` calls.  In fact, we'll need to get rid of ALL of them before we are done with Angular.  And I'm hoping Ed will agree we should kill the autoclassify calls in this class, which will delete even more of them.  :-)  

So, if you're cool with it, let's table the idea until my ``selectedJob`` PR?